### ISSUE TITLE
Run_ eval is not working

### ISSUE THREAD

**Issue by dynamic-tyrant** [zjunlp/AutoAct#15]
Why hasn't anything happened when running Python Self-Planning/Group-Planning/run_ eval. py

**Comment by GoooDte** [comment]
Hello, can you provide a screenshot of your operating shell or the messages of the reporting error?

**Comment by dynamic-tyrant** [comment]
------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Shuofei ***@***.***&gt;; 
发送时间: 2024年12月26日(星期四) 下午5:22
收件人: ***@***.***&gt;; 
抄送: ***@***.***&gt;; ***@***.***&gt;; 
主题: Re: [zjunlp/AutoAct] Run_ eval is not working (Issue #15)














 
Hello, can you provide a screenshot of your operating shell or the messages of the reporting error?

**Comment by dynamic-tyrant** [comment]
![图片](https://github.com/user-attachments/assets/cb601b42-27b3-45ed-8557-b2e0698e7558)

**Comment by GoooDte** [comment]
Do you have a model deployed for evaluation?

**Comment by dynamic-tyrant** [comment]
Yes, I have already launched the model using Fastchat. Do we need to set the path of the corresponding file after training in the parameters plan_agent, tool_agent, and reflect_agent? However, your source code does not have a corresponding parameter for tool_agent, but instead has an additional parameter called action_agent

**Comment by GoooDte** [comment]
You need to deploy the three agents at the same time with different names.

**Comment by GoooDte** [comment]
please refer to https://github.com/zjunlp/AutoAct/blob/main/Scripts/model_bash/multi_model.sh

**Comment by dynamic-tyrant** [comment]
Now I have another issue where an error occurs while running Agent training, [rank0]: AttributeError: module 'deepspeed' has no attribute 'is_deepspeed_zero3_enabled'

**Comment by dynamic-tyrant** [comment]
Also, may I ask which version of DeepSpeed you have installed

**Comment by dynamic-tyrant** [comment]
> please refer to https://github.com/zjunlp/AutoAct/blob/main/Scripts/model_bash/multi_model.sh

The script was successfully executed, but run_ eval. py still doesn't work

**Comment by GoooDte** [comment]
sincerely recommend you to have a debug and see why there is no output

**Comment by dynamic-tyrant** [comment]
I have successfully run run_ eval. py and output the result, but returned some errors:
Thought: Hmm, I think I should search for information about Malcolm Smith on the Internet.
*********temperature0.5***********************
*********temperature0.5***********************
Action: BingSearch[Malcolm Smith]
<Response [401]>
Observation: bingSearch error, please try again.
*********temperature0.5***********************
Thought: Oh, I see. Let me try again with a different search query.
*********temperature0.5***********************
*********temperature0.5***********************
Action: BingSearch[Malcolm Smith]
<Response [401]>

**Comment by dynamic-tyrant** [comment]
> It seems that you need to set a bing search key.

Where do I need to set the Bing search key? You did not indicate it in the readme

**Comment by GoooDte** [comment]
https://www.microsoft.com/en-us/bing/apis/bing-web-search-api

**Comment by GoooDte** [comment]
It needs money, about 100 dollars for the whole experiment.

**Comment by zxlzr** [comment]
hi, do you have any further questions?

**Comment by dynamic-tyrant** [comment]
I have completed all the steps, but the effect of using the llama2-7b model is not good. In addition, during Lora training, the loss value has been hovering at a certain level and has not decreased, sometimes resulting in errors: [rank1]: Exception: Current loss scale already at minimum - cannot decrease scale anymore. Exiting run.

**Comment by GoooDte** [comment]
May I ask which value your loss hovered at and what your training data looks like?

**Comment by dynamic-tyrant** [comment]
I used the data you published on Google Drive for training, and here are some of the training loss results I have extracted:
{'loss': 0.4506, 'learning_rate': 9.43072021389003e-05, 'epoch': 0.21}                                                                                                                                                     
{'loss': 0.8168, 'learning_rate': 9.38524223653626e-05, 'epoch': 0.22}                                                                                                                                                     
{'loss': 0.2905, 'learning_rate': 9.338135320708911e-05, 'epoch': 0.22}                                                                                                                                                    
{'loss': 0.4359, 'learning_rate': 9.289416964704185e-05, 'epoch': 0.23}                                                                                                                                                    
{'loss': 0.9769, 'learning_rate': 9.239105265402525e-05, 'epoch': 0.24}                                                                                                                                                    
{'loss': 0.8658, 'learning_rate': 9.187218911546362e-05, 'epoch': 0.24}

**Comment by GoooDte** [comment]
The training seems to do not converge, may I see your training configs?

**Comment by dynamic-tyrant** [comment]
> The training seems to not converge, may I see your training configs?

for agent in plan action reflect
do
echo "####################"
echo $agent
echo "####################"
CUDA_VISIBLE_DEVICES=0,1 deepspeed Self_Plan/Train/train_lora.py \
--model_name_or_path /home/q/llama2_7b_hf \
    --lora_r 8 \
    --lora_alpha 16 \
    --lora_dropout 0.05 \
    --data_path Self_Plan/Traj_Syn/output/data_$agent.json \
    --output_dir Self_Plan/Train/lora/HotpotQA/7b-$agent-5-epoch \
    --num_train_epochs 5 \
    --per_device_train_batch_size 2 \
    --per_device_eval_batch_size 1 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 10000 \
    --save_total_limit 1 \
    --learning_rate 1e-4 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --fp16 True \
    --model_max_length 4096 \
    --gradient_checkpointing True \
    --q_lora False \
    --deepspeed Self_Plan/Train/deepspeed_config_s3.json \
    --resume_from_checkpoint False
done

**Comment by GoooDte** [comment]
We recommend you use the chat version model (llama-2-chat-7b) for training instead of just the pre-trained model, because the former has been well-trained on instructions, so the loss will be more stable.

**Comment by zxlzr** [comment]
hi, do you have any further questions？

**Comment by dynamic-tyrant** [comment]
I am using the untrained version (llama-2-chat-7b), but the accuracy rate is only 15%. The accuracy of llama-2-chat-13b is only 28%

**Comment by Rolnand** [comment]
Autoact is a agent-tuning method, so you need to train the model. See the script in the Self-Differentiation of README

**Comment by dynamic-tyrant** [comment]
Yes, I also trained the model, but the effect is just not good

**Comment by Rolnand** [comment]
Have you deployed the differentiation model of lora weighted after training? It seems like untrained model deployed because the results is  like the results of BOLAA in llama-2-chat-13b

**Comment by dynamic-tyrant** [comment]
I used the fastchat method to load the Lora trained model

**Comment by Rolnand** [comment]
Could you show your fastchat script

**Comment by dynamic-tyrant** [comment]
CUDA_VISIBLE_DEVICES=0,1 PEFT_SHARE_BASE_WEIGHTS=true python3 -m fastchat.serve.multi_model_worker \
    --port 31022 --worker http://localhost:31022 \
    --host localhost \
    --model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-plan-5-epoch \
    --model-names "plan" \
    --model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-action-5-epoch \
    --model-names "action" \
    --model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-reflect-5-epoch \
    --model-names "reflect" \
    --max-gpu-memory 31Gib \
    --dtype float16 \
--num-gpus 2

**Comment by Rolnand** [comment]
try
CUDA_VISIBLE_DEVICES=0,1 PEFT_SHARE_BASE_WEIGHTS=true python3 -m fastchat.serve.multi_model_worker
--port 31022 --worker http://localhost:31022/
--host localhost
--model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-plan-5-epoch
--model-names "plan_peft"
--model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-action-5-epoch
--model-names "action_peft"
--model-path /home/q/AutoAct-main/Self_Plan/Train/lora/HotpotQA/7b-reflect-5-epoch
--model-names "reflect_peft"
--max-gpu-memory 31Gib
--dtype float16
--num-gpus 2

**Comment by zxlzr** [comment]
have you solved your issue yet?

**Comment by Rolnand** [comment]
We saved the output trajectory at that time, you can compare your trajectory to see where the gap is.(Maybe BingSearch Exceeding the limit or Unstable model calls) 
Please leave your email so we can send you the trajectory file.

**Comment by dynamic-tyrant** [comment]
> We saved the output trajectory at that time, you can compare your trajectory to see where the gap is.(Maybe BingSearch Exceeding the limit or Unstable model calls) Please leave your email so we can send you the trajectory file.

1369485614@qq.com
